[SwiftUI] Support more exported content types from WebPage

rr-codes opened the pull request:
#### 248632b107f495370f87b9a8f658366f742e4dce
<pre>
[SwiftUI] Support more exported content types from WebPage
<a href="https://bugs.webkit.org/show_bug.cgi?id=296233">https://bugs.webkit.org/show_bug.cgi?id=296233</a>
<a href="https://rdar.apple.com/156225187">rdar://156225187</a>

Reviewed by Wenson Hsieh.

Add support for exporting the contents of a WebPage as plain text and rich text

* Source/WebKit/UIProcess/API/Swift/WebPage+Transferable.swift:
(transferRepresentation):

- Refine the `.image` export representation to `.png`, which is more specific and
usually more useful (for example, it can infer a file extension).

- Expose additional representations for `.flatRTFD`, `.rtf`, and `.utf8PlainText`

(WebPage.plainTextRepresentation):

- Use existing WebKit functionality to get the contents of all frames as a String, a
nd use its data for the export.

(WebPage.richTextRepresentation(_:)):

- Use existing WebKit functionality to get the contents of the page as an NSAttributedString
with a specified document type, and then use its data for the export along with the relevant
document attributes.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/Foundation+Extras.swift:
(AsyncSequence.wait(_:)):
(URL.aboutBlank): Deleted.

- Fix some formatting
- Replace the Array extension with the RangeReplaceableCollection extension since having both
was redundant
- Add a convenience extension function to AsyncSequence to wait until the sequence is finished.
- No need for an `aboutBlank` property anymore now that the API supports default values for base URLs.

* Tools/TestWebKitAPI/Tests/WebKit Swift/SwiftUI+Extras.swift: Added.

- Make it easier to convert between an AttributedString&apos;s CocoaAttributes&apos; fonts values and SwiftUI&apos;s
`Font` type.

* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigationTests.swift:
(WebPageNavigationTests.basicNavigationProducesExpectedNavigationEvents):
(WebPageNavigationTests.failedNavigationProducesExpectedNavigationError):
(WebPageNavigationTests.explicitlyStopLoadingProgrammaticNavigation):
(WebPageNavigationTests.stopLoadingProgrammaticNavigationViaTaskCancellation):
(WebPageNavigationTests.failedNavigationWithWebContentProcessTerminated):
(WebPageNavigationTests.navigationProceedsAfterDiscardingNavigationStream):

- Use the updated convenience APIs to simplify some of the code
- Fix formatting/linting issues

* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageTransferableTests.swift:
(WebPageTransferableTests.transferableContentTypes):
(WebPageTransferableTests.exportToPDFWithFullContent):
(WebPageTransferableTests.exportToPDFWithSubRegion):
(WebPageTransferableTests.exportToURL):
(WebPageTransferableTests.exportToImage(_:)):
(WebPageTransferableTests.exportToPlainText(_:)):
(WebPageTransferableTests.exportToRichTextWithDocumentAttributes(_:)):
(WebPageTransferableTests.exportToImage): Deleted.

Add tests for all the new exportable content types.

Canonical link: <a href="https://commits.webkit.org/297658@main">https://commits.webkit.org/297658@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cdde402284fad211c6572ead7f06e2f4a191837f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112433 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32165 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22642 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118512 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62836 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114395 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32817 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40728 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85409 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115380 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26192 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101155 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; 1 api test failed or timed out; re-run-api-tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65840 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25488 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19290 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62357 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95579 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19365 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121838 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39507 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29417 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/94217 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39889 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97392 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94042 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24046 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39289 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17090 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39395 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44883 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39030 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42367 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40773 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -743,9 +744,9 @@
-		A176D24D2D6D2A4E00C63915 /* WebContentExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2492D6D295400C63915 /* WebContentExtension.appex */; platformFilter = ios; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		A176D24B2D6D2A4E00C63915 /* GPUExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2472D6D295400C63915 /* GPUExtension.appex */; platformFilters = (ios, ); settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		A176D24C2D6D2A4E00C63915 /* NetworkingExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2482D6D295400C63915 /* NetworkingExtension.appex */; platformFilters = (ios, ); settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		A176D24D2D6D2A4E00C63915 /* WebContentExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2492D6D295400C63915 /* WebContentExtension.appex */; platformFilters = (ios, ); settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
```
Are these intentional?

rr-codes replied:
ah, not really, but I figured Xcode wanted it so it can't hurt? I can put it back to what it was before if that's better

whsieh replied:
I see, either way is fine with me (as long as other versions of Xcode don't try to force `platformFilter = ios` again)

rr-codes replied:
ah, true. ill just put it back

rr-codes committed:
[SwiftUI] Support more exported content types from WebPage
https://bugs.webkit.org/show_bug.cgi?id=296233
rdar://156225187

Reviewed by Wenson Hsieh.

Add support for exporting the contents of a WebPage as plain text and rich text

* Source/WebKit/UIProcess/API/Swift/WebPage+Transferable.swift:
(transferRepresentation):

- Refine the `.image` export representation to `.png`, which is more specific and
usually more useful (for example, it can infer a file extension).

- Expose additional representations for `.flatRTFD`, `.rtf`, and `.utf8PlainText`

(WebPage.plainTextRepresentation):

- Use existing WebKit functionality to get the contents of all frames as a String, a
nd use its data for the export.

(WebPage.richTextRepresentation(_:)):

- Use existing WebKit functionality to get the contents of the page as an NSAttributedString
with a specified document type, and then use its data for the export along with the relevant
document attributes.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/Foundation+Extras.swift:
(AsyncSequence.wait(_:)):
(URL.aboutBlank): Deleted.

- Fix some formatting
- Replace the Array extension with the RangeReplaceableCollection extension since having both
was redundant
- Add a convenience extension function to AsyncSequence to wait until the sequence is finished.
- No need for an `aboutBlank` property anymore now that the API supports default values for base URLs.

* Tools/TestWebKitAPI/Tests/WebKit Swift/SwiftUI+Extras.swift: Added.

- Make it easier to convert between an AttributedString's CocoaAttributes' fonts values and SwiftUI's
`Font` type.

* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigationTests.swift:
(WebPageNavigationTests.basicNavigationProducesExpectedNavigationEvents):
(WebPageNavigationTests.failedNavigationProducesExpectedNavigationError):
(WebPageNavigationTests.explicitlyStopLoadingProgrammaticNavigation):
(WebPageNavigationTests.stopLoadingProgrammaticNavigationViaTaskCancellation):
(WebPageNavigationTests.failedNavigationWithWebContentProcessTerminated):
(WebPageNavigationTests.navigationProceedsAfterDiscardingNavigationStream):

- Use the updated convenience APIs to simplify some of the code
- Fix formatting/linting issues

* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageTransferableTests.swift:
(WebPageTransferableTests.transferableContentTypes):
(WebPageTransferableTests.exportToPDFWithFullContent):
(WebPageTransferableTests.exportToPDFWithSubRegion):
(WebPageTransferableTests.exportToURL):
(WebPageTransferableTests.exportToImage(_:)):
(WebPageTransferableTests.exportToPlainText(_:)):
(WebPageTransferableTests.exportToRichTextWithDocumentAttributes(_:)):
(WebPageTransferableTests.exportToImage): Deleted.

Add tests for all the new exportable content types.

Canonical link: https://commits.webkit.org/297658@main
